common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/WorkType.scala

```diff
@@ -1,5 +1,160 @@
 package uk.ac.wellcome.models.work.internal
 
-case class WorkType(id: String,
-                    label: String,
-                    ontologyType: String = "WorkType")
+import enumeratum.{Enum, EnumEntry}
+import io.circe.{Decoder, Encoder, Json}
+
+sealed trait WorkType extends EnumEntry {
+  val id: String
+  val label: String
+}
+
+object WorkType extends Enum[WorkType] {
+  val values = findValues
+
+  implicit val workTypeEncoder: Encoder[WorkType] = Encoder.instance[WorkType] {
+    workType =>
+      Json.obj(
+        ("id", Json.fromString(workType.id)),
+        ("label", Json.fromString(workType.label))
+      )
+  }
+
+  implicit val workTypeDecoder: Decoder[WorkType] = Decoder.instance[WorkType] {
+    cursor =>
+      for {
+        id <- cursor.downField("id").as[String]
+      } yield {
+        fromCode(id).getOrElse(throw new Exception(s"Invalid WorkType id: $id"))
+      }
+  }
+
+  def fromCode(id: String): Option[WorkType] = {
+    values.find(workType => workType.id == id)
+  }
+
+  sealed trait Unlinked extends WorkType
```

Comment:
So this is both an Enum _and_ an ADT?

Reply:
Just speculating here, but it's likely IMO that the classic concept of an enum was only invented due to lack of ADT support in certain languages, since ADTs give the same functionality but in more type safe manner (i.e. with pattern matching). Worth mentioning that an `enum` [in Rust](https://doc.rust-lang.org/book/ch06-01-defining-an-enum.html) provides pretty much the same thing as `sealed trait` / `case class` in Scala.